File: codes/golang/19.删除链表的倒数第N个节点.go

// 主函数
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	// 虚拟头结点
	dummy := &ListNode{Val: -1, Next: head}
	// 删除倒数第 n 个，要先找倒数第 n + 1 个节点
	x := findFromEnd(dummy, n+1)
	// 删掉倒数第 n 个节点
	x.Next = x.Next.Next
	return dummy.Next
}

// 返回链表的倒数第 k 个节点
func findFromEnd(head *ListNode, k int) *ListNode {
	p1, p2 := head, head
	// p1 先走 k 步
	for i := 0; i < k; i++ {
		p1 = p1.Next
	}
	// p1 和 p2 同时走 n - k 步
	for p1 != nil {
		p1 = p1.Next
		p2 = p2.Next
	}
	// p2 现在指向第 n - k 个节点
	return p2
}

File: codes/rust/19.删除链表的倒数第N个节点.rs
/*
 * @lc app=leetcode.cn id=19 lang=rust
 *
 * [19] 删除链表的倒数第 N 个结点
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        let mut dummy = Some(Box::new(ListNode::new(0)));
        dummy.as_mut().unwrap().next = head;
        let mut first = dummy.clone();
        let mut second = dummy.as_mut();

        for _ in 0..=n {
            first = first.unwrap().next.clone();
        }

        while first.is_some() {
            first = first.unwrap().next.clone();
            second = second.unwrap().next.as_mut();
        }

        let mut node_to_remove = second.as_mut().unwrap().next.take();
        second.as_mut().unwrap().next = node_to_remove.as_mut().unwrap().next.take();

        dummy.unwrap().next
    }
}
// @lc code=end
